document/encoding/msgpack/encoding.go

```diff
@@ -0,0 +1,306 @@
+package msgpack
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/genjidb/genji/document"
+	"github.com/vmihailenco/msgpack/v5"
+	"github.com/vmihailenco/msgpack/v5/codes"
+)
+
+// EncodeDocument takes a document and encodes it
+// in MessagePack.
+func EncodeDocument(d document.Document) ([]byte, error) {
+	if ec, ok := d.(EncodedDocument); ok {
+		return ec, nil
+	}
+
+	var buf bytes.Buffer
+	enc := NewEncoder(&buf)
+	defer enc.Close()
+
+	err := enc.EncodeDocument(d)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// DecodeDocument takes a byte slice and returns a lazily decoded document.
+// If buf is malformed, an error will be returned when calling one of the document method.
+func DecodeDocument(buf []byte) document.Document {
+	return EncodedDocument(buf)
+}
+
+// EncodeValue encodes any value to MessagePack.
+func EncodeValue(w io.Writer, v document.Value) error {
+	enc := NewEncoder(w)
+	defer enc.Close()
+
+	return enc.EncodeValue(v)
+}
+
+// An EncodedDocument implements the document.Document
+// interface on top of an encoded representation of a
+// document.
+// It is useful for avoiding decoding the entire document when
+// only a few fields are needed.
```

Comment:
Not really, we may need only a few fields, not only one.
Example:

```sql
..... WHERE a = 10 AND b = 11
```

The expression `a = 10 AND b = 11` will call `GetByField("a")` and `GetByField("b")`, which means that it will need to decode two fields, but not all of them.